### ISSUE TITLE
MPI with multiple fields and injection generates bad code

### ISSUE THREAD

**Issue by mikeando** [devitocodes/devito#1119]
I've cut down some Born style code which does roughly:

u<sub>tt</sub> = ∇<sup>2</sup>u
u = src.inject
U<sub>tt</sub> = u<sub>tt</sub> + ∇<sup>2</sup>U

The generated C code includes a `haloupdate0(U_vec,comm,nb,t0,nthreads_nonaffine);` inside the core of the `bf0` loop, which is "bad". Also it places a `#pragma omp simd aligned(U,u:32)` when there is no `U` variable in the loop.

The python generating this is:

```python
import numpy as np
from sympy import Symbol

import mpi4py

mpi4py.rc.finalize = False
mpi4py.rc.initialize = False
from mpi4py import MPI

MPI.Init_thread()

from devito import configuration

configuration["mpi"] = "full"
configuration["develop-mode"] = False
configuration["openmp"] = True

from devito import (
    SparseTimeFunction,
    Operator,
    TimeFunction,
    Grid,
    Constant,
    SpaceDimension,
    SubDimension,
    Eq,
    solve,
)

spacedims = ["x", "y", "z"]
spacings = [Constant(name="h_%s" % d, dtype=np.float32) for d in spacedims]
dimensions = tuple(SpaceDimension(d, spacing=s) for d, s in zip(spacedims, spacings))

internal_dims = tuple(
    [
        SubDimension.middle(
            name=d.name + "i", parent=d, thickness_left=2, thickness_right=2
        )
        for d in dimensions
    ]
)

internal_dim_subs = {d: di for (d, di) in zip(dimensions, internal_dims)}

space_order = 16

grid = Grid(shape=(20, 20, 20), dimensions=dimensions, comm=MPI.COMM_SELF)

u = TimeFunction(
    name="u",
    grid=grid,
    time_order=2,
    space_order=(space_order, space_order // 2, space_order // 2),
)

U = TimeFunction(
    name="U",
    grid=grid,
    time_order=2,
    space_order=(space_order, space_order // 2, space_order // 2),
)

wave_rhs = Symbol("wave_rhs")


def wave_step(_u):
    return Eq(_u.forward, solve(Eq(_u.dt2, wave_rhs), _u.forward))


wave_step_u = wave_step(u).xreplace({wave_rhs: u.laplace}).xreplace(internal_dim_subs)
wave_step_U = (
    wave_step(U).xreplace({wave_rhs: 2 * u.dt2 + U.laplace}).xreplace(internal_dim_subs)
)

coords = np.array([(0.5, 0.5, 0.5), (0.5, 2.5, 0.5), (2.5, 0.5, 0.5), (2.5, 2.5, 0.5)])
src = SparseTimeFunction(
    name="src", grid=grid, nt=3, npoint=len(coords), coordinates=coords
)
wave_src_inject = src.inject(field=u.forward, expr=src * (grid.time_dim.spacing ** 2))

print(wave_src_inject)
actions_fwd_downup = [wave_step_u] + wave_src_inject + [wave_step_U]

op_fwd_downup = Operator(
    actions_fwd_downup, name="BornIso3DFwdDownUp", dse="aggressive", dle="advanced"
)

print(op_fwd_downup.ccode)
```

And the relevant part of the output C code is:

```C
//...
void bf0(struct dataobj *restrict U_vec, const float dt, const float h_x, const float h_y, const float h_z, struct dataobj *restrict u_vec, MPI_Comm comm, struct neighborhood * nb, const int t0, const int t1, const int t2, const int xi0_blk0_M, const int xi0_blk0_m, const int xi0_blk0_size, const int yi0_blk0_M, const int yi0_blk0_m, const int yi0_blk0_size, const int z_M, const int z_m, const int zi_ltkn, const int zi_rtkn, const int nthreads_nonaffine, const int nthreads)
{
  float (*restrict u)[u_vec->size[1]][u_vec->size[2]][u_vec->size[3]] __attribute__ ((aligned (64))) = (float (*)[u_vec->size[1]][u_vec->size[2]][u_vec->size[3]]) u_vec->data;
  if (xi0_blk0_size == 0)
  {
    return;
  }
  #pragma omp parallel num_threads(nthreads)
  {
    #pragma omp for collapse(1) schedule(dynamic,1)
    for (int xi0_blk0 = xi0_blk0_m; xi0_blk0 <= xi0_blk0_M; xi0_blk0 += xi0_blk0_size)
    {
      #pragma omp simd aligned(U,u:32)
      for (int yi0_blk0 = yi0_blk0_m; yi0_blk0 <= yi0_blk0_M; yi0_blk0 += yi0_blk0_size)
      {
        for (int xi = xi0_blk0; xi <= xi0_blk0 + xi0_blk0_size - 1; xi += 1)
        {
          for (int yi = yi0_blk0; yi <= yi0_blk0 + yi0_blk0_size - 1; yi += 1)
          {
            for (int zi = z_m + zi_ltkn; zi <= z_M - zi_rtkn; zi += 1)
            {
              haloupdate0(U_vec,comm,nb,t0,nthreads_nonaffine);
              float r6 = -3.0548441F*u[t0][xi + 8][yi + 8][zi + 8];
              float r4 = (r6 - 2.42812743e-6F*(u[t0][xi + 8][yi + 8][zi] + u[t0][xi + 8][yi + 8][zi + 16]) + 5.07429079e-5F*(u[t0][xi + 8][yi + 8][zi + 1] + u[t0][xi + 8][yi + 8][zi + 15]) - 5.18000518e-4F*(u[t0][xi + 8][yi + 8][zi + 2] + u[t0][xi + 8][yi + 8][zi + 14]) + 3.48096348e-3F*(u[t0][xi + 8][yi + 8][zi + 3] + u[t0][xi + 8][yi + 8][zi + 13]) - 1.76767677e-2F*(u[t0][xi + 8][yi + 8][zi + 4] + u[t0][xi + 8][yi + 8][zi + 12]) + 7.54208754e-2F*(u[t0][xi + 8][yi + 8][zi + 5] + u[t0][xi + 8][yi + 8][zi + 11]) - 3.11111111e-1F*(u[t0][xi + 8][yi + 8][zi + 6] + u[t0][xi + 8][yi + 8][zi + 10]) + 1.77777778F*(u[t0][xi + 8][yi + 8][zi + 7] + u[t0][xi + 8][yi + 8][zi + 9]))/((h_z*h_z)) + (r6 - 2.42812743e-6F*(u[t0][xi + 8][yi][zi + 8] + u[t0][xi + 8][yi + 16][zi + 8]) + 5.07429079e-5F*(u[t0][xi + 8][yi + 1][zi + 8] + u[t0][xi + 8][yi + 15][zi + 8]) - 5.18000518e-4F*(u[t0][xi + 8][yi + 2][zi + 8] + u[t0][xi + 8][yi + 14][zi + 8]) + 3.48096348e-3F*(u[t0][xi + 8][yi + 3][zi + 8] + u[t0][xi + 8][yi + 13][zi + 8]) - 1.76767677e-2F*(u[t0][xi + 8][yi + 4][zi + 8] + u[t0][xi + 8][yi + 12][zi + 8]) + 7.54208754e-2F*(u[t0][xi + 8][yi + 5][zi + 8] + u[t0][xi + 8][yi + 11][zi + 8]) - 3.11111111e-1F*(u[t0][xi + 8][yi + 6][zi + 8] + u[t0][xi + 8][yi + 10][zi + 8]) + 1.77777778F*(u[t0][xi + 8][yi + 7][zi + 8] + u[t0][xi + 8][yi + 9][zi + 8]))/((h_y*h_y)) + (r6 - 2.42812743e-6F*(u[t0][xi][yi + 8][zi + 8] + u[t0][xi + 16][yi + 8][zi + 8]) + 5.07429079e-5F*(u[t0][xi + 1][yi + 8][zi + 8] + u[t0][xi + 15][yi + 8][zi + 8]) - 5.18000518e-4F*(u[t0][xi + 2][yi + 8][zi + 8] + u[t0][xi + 14][yi + 8][zi + 8]) + 3.48096348e-3F*(u[t0][xi + 3][yi + 8][zi + 8] + u[t0][xi + 13][yi + 8][zi + 8]) - 1.76767677e-2F*(u[t0][xi + 4][yi + 8][zi + 8] + u[t0][xi + 12][yi + 8][zi + 8]) + 7.54208754e-2F*(u[t0][xi + 5][yi + 8][zi + 8] + u[t0][xi + 11][yi + 8][zi + 8]) - 3.11111111e-1F*(u[t0][xi + 6][yi + 8][zi + 8] + u[t0][xi + 10][yi + 8][zi + 8]) + 1.77777778F*(u[t0][xi + 7][yi + 8][zi + 8] + u[t0][xi + 9][yi + 8][zi + 8]))/((h_x*h_x));
              float r5 = 2.0F*u[t0][xi + 8][yi + 8][zi + 8] - 1.0F*u[t2][xi + 8][yi + 8][zi + 8];
              u[t1][xi + 8][yi + 8][zi + 8] = 1.0F*r4*(dt*dt) + r5;
            }
          }
        }
      }
    }
  }
}
//...
```

**Comment by FabioLuporini** [comment]
fixed by #1115

**Comment by FabioLuporini** [comment]
fixed by #1115
